pig/src/org/partiql/pig/domain/parser/TypeDomainParser.kt

```diff
@@ -15,275 +15,350 @@
 
 package org.partiql.pig.domain.parser
 
-import com.amazon.ionelement.api.*
 import com.amazon.ion.IonReader
 import com.amazon.ion.system.IonReaderBuilder
-import org.partiql.pig.domain.model.*
+import com.amazon.ionelement.api.AnyElement
+import com.amazon.ionelement.api.IonElement
+import com.amazon.ionelement.api.IonElementException
+import com.amazon.ionelement.api.IonElementLoaderOptions
+import com.amazon.ionelement.api.IonLocation
+import com.amazon.ionelement.api.MetaContainer
+import com.amazon.ionelement.api.SexpElement
+import com.amazon.ionelement.api.SymbolElement
+import com.amazon.ionelement.api.createIonElementLoader
+import com.amazon.ionelement.api.emptyMetaContainer
+import com.amazon.ionelement.api.head
+import com.amazon.ionelement.api.location
+import com.amazon.ionelement.api.metaContainerOf
+import com.amazon.ionelement.api.tag
+import com.amazon.ionelement.api.tail
+import org.partiql.pig.domain.model.Arity
+import org.partiql.pig.domain.model.DataType
+import org.partiql.pig.domain.model.NamedElement
+import org.partiql.pig.domain.model.PermutedDomain
+import org.partiql.pig.domain.model.PermutedSum
+import org.partiql.pig.domain.model.Statement
+import org.partiql.pig.domain.model.Transform
+import org.partiql.pig.domain.model.TupleType
+import org.partiql.pig.domain.model.TypeDomain
+import org.partiql.pig.domain.model.TypeRef
+import org.partiql.pig.domain.model.TypeUniverse
+import org.partiql.pig.errors.ErrorContext
+import java.io.File
+import java.io.FileNotFoundException
+import java.io.InputStream
+import java.util.Stack
+
+
+fun parseTypeUniverseFile(path: String): TypeUniverse {
+    val parser = Parser(FILE_SYSTEM_SOURCE)
+    return parser.parseTypeUniverse(path)
+}
+
+internal class Parser(
+    val inputSource: InputSource
+) {
+    /** This contains every file the parser has "seen" and is used detect and prevent import cycles. */
+    private val parseHistory = HashSet<String>()
+    private val qualifedSourceStack = Stack<String>().apply { push(".") }
```

Comment:
This comment seems a little incongruent with what this PR actually does, which is to allow the a type universe to be defined across multiple files.  There is still only ever one (unnamed) type universe--it can now just reside in multiple files.  Thus, I've included the filename here instead.

Reply:
Incidentally, pushing `"."` to the stack here doesn't make sense to me anymore and anyway wasn't actually needed.  Removed.